optional_plugins/varianter_cit/avocado_varianter_cit/__init__.py

```diff
@@ -0,0 +1,331 @@
+            for key, val in variant.items():
+                variant_tree_nodes.append(TreeNode(key, {key: val}))
+
+            yield {"variant_id": vid,
+                   "variant": variant_tree_nodes,
+                   "paths": self.parameter_path}
+
+    def __len__(self):
+        return sum(1 for _ in self)
+
+    def update_defaults(self, defaults):
+        pass
+
+    def to_str(self, summary, variants, **kwargs):
+        """
+        Return human readable representation
+
+        The summary/variants accepts verbosity where 0 means silent and
+        maximum is up to the plugin.
+
+        :param summary: How verbose summary to output (int)
+        :param variants: How verbose list of variants to output (int)
+        :param kwargs: Other free-form arguments
+        :rtype: str
+        """
+        if not self.variants:
+            return ""
+
+        out = []
+        verbose = variants > 1
+        out.append("CIT Variants (%i):" % len(self))
+        for variant in self:
+            out.append('%sVariant %s:    %s' % ('\n' if verbose else '',
+                                                variant["variant_id"],
+                                                self.parameter_path))
+            if not verbose:
+                continue
+            env = set()
+            for node in variant["variant"]:
+                for key, value in iteritems(node.environment):
+                    origin = node.environment.origin[key].path
+                    env.add(("%s:%s" % (origin, key), str(value)))
+            if not env:
+                return out
+            fmt = '    %%-%ds => %%s' % max([len(_[0]) for _ in env])
+            for record in sorted(env):
+                out.append(fmt % record)
+
+        return "\n".join(out)
+
+
+class Cit(object):
+
+    MATRIX_ROW_SIZE = 20
+    MAX_INTERACTIONS = 15
+
+    hash_table = {}
+
+    def __init__(self, parameters, constraints, order):
+        self.parameters = parameters
+        self.data = [len(param[1]) for param in self.parameters]
+        # Constraints are not implemented in this example
+        self.constraints = constraints
+        self.order = order
+
+    def combine(self):
+        """
+        Computes the combination of parameters
+
+        :returns: headers (list of parameters keys) and combinations (list of
+                  dictionaries. Each dictionary represents a combination of
+                  parameters.
+        :rtype: tuple
+        """
+        self.create_interaction_hash_table()
+        final_list = self.create_final_list()
+
+        headers = [item[0] for item in self.parameters]
+        result = [[self.parameters[i][1][combination[i]] for i in range(len(combination))]for combination in final_list]
+        combinations = []
+        for combination in result:
+            combinations.append(dict(zip(headers, combination)))
+
+        return headers, combinations
+
+    def create_interaction_hash_table(self):
```

Comment:
I think we always want to run this and only once. Shouldn't it be part of the `__init__`? If not we expect them to change so we should clean the previous `self.hash_table`.